pkg/operator/csidriveroperator/crcontroller.go

```diff
@@ -240,6 +221,61 @@ func (c *CSIDriverOperatorCRController) applyClusterCSIDriver(requiredOriginal *
 	return actual, true, err
 }
 
+func (c *CSIDriverOperatorCRController) syncConditions(conditions []operatorapi.OperatorCondition, updatefn v1helpers.UpdateStatusFunc) error {
+	var availableCnd operatorapi.OperatorCondition
+	disabled, msg := c.hasDisabledCondition(conditions)
+	if disabled && c.optional {
+		// The driver can't be running. Mark the operator as Available, but with an extra message.
+		availableCnd.Status = operatorapi.ConditionTrue
+		availableCnd.Reason = "DriverDisabled"
+		availableCnd.Message = fmt.Sprintf("CSI driver for %s is disabled: %s", c.name, msg)
+	} else {
+		// The driver should be running, copy conditions from the CR
+		availableCnd = status.UnionCondition(operatorapi.OperatorStatusTypeAvailable, operatorapi.ConditionTrue, nil, conditions...)
+		if availableCnd.Status == operatorapi.ConditionUnknown {
+			availableCnd.Status = operatorapi.ConditionFalse
+			availableCnd.Reason = "WaitForOperator"
+			availableCnd.Message = fmt.Sprintf("Waiting for %s operator to report status", c.name)
+		}
+	}
+	availableCnd.Type = c.crConditionName(operatorapi.OperatorStatusTypeAvailable)
+
+	progressingCnd := status.UnionCondition(operatorapi.OperatorStatusTypeProgressing, operatorapi.ConditionFalse, nil, conditions...)
+	progressingCnd.Type = c.crConditionName(operatorapi.OperatorStatusTypeProgressing)
+	if progressingCnd.Status == operatorapi.ConditionUnknown {
+		if disabled && c.optional {
+			progressingCnd.Status = operatorapi.ConditionFalse
+		} else {
+			progressingCnd.Status = operatorapi.ConditionTrue
+			progressingCnd.Reason = "WaitForOperator"
+			progressingCnd.Message = fmt.Sprintf("Waiting for %s operator to report status", c.name)
+		}
+	}
+
+	degradedCnd := status.UnionCondition(operatorapi.OperatorStatusTypeDegraded, operatorapi.ConditionFalse, nil, conditions...)
+	degradedCnd.Type = c.crConditionName(operatorapi.OperatorStatusTypeDegraded)
+	if degradedCnd.Status == operatorapi.ConditionUnknown {
+		degradedCnd.Status = operatorapi.ConditionFalse
+	}
```

Comment:
Can we have an operator that's disabled&optional but is also progressing or degraded?

Doesn't sound like we want to handle that, but just checking...

Reply:
Disabled & degraded can happen:

1. Manila operator checks there is no Manila in OpenStack and sets `Disabled: true`
2. On regular re-sync, Manila fails to talk to OpenStack and reports `Degraded: true`. This keeps `Disabled: true` (at least I think so)

With the current code, CSO will report `Degraded: true`.

Disabled & Progressing is IMO impossible, but CSO should be robust enough to handle that.